p-token/src/entrypoint.rs

```diff
@@ -14,218 +13,245 @@ default_panic_handler!();
 
 /// Process an instruction.
 ///
+/// In the first stage, the entrypoint checks the discriminator of the instruction data
+/// to determine whether the instruction is a "batch" instruction or a "regular" instruction.
+/// This avoids nesting of "batch" instructions, since it is not sound to have a "batch"
+/// instruction inside another "batch" instruction.
+#[inline(always)]
+pub fn process_instruction(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    let [discriminator, remaining @ ..] = instruction_data else {
+        return Err(ProgramError::InvalidInstructionData);
+    };
+
+    if *discriminator == 255 {
+        // 255 - Batch
+        #[cfg(feature = "logging")]
+        pinocchio::msg!("Instruction: Batch");
+
+        return process_batch(accounts, remaining);
+    }
+
+    inner_process_instruction(accounts, instruction_data)
+}
+
+/// Process a "regular" instruction.
+///
 /// The processor of the token program is divided into two parts to reduce the overhead
 /// of having a large `match` statement. The first part of the processor handles the
 /// most common instructions, while the second part handles the remaining instructions.
+///
 /// The rationale is to reduce the overhead of making multiple comparisons for popular
 /// instructions.
 ///
-/// Instructions on the first part of the processor:
+/// Instructions on the first part of the inner processor:
 ///
-/// - `0`: `InitializeMint`
-/// - `3`:  `Transfer`
-/// - `7`:  `MintTo`
-/// - `9`:  `CloseAccount`
+/// -  `0`: `InitializeMint`
+/// -  `1`: `InitializeAccount`
+/// -  `3`: `Transfer`
+/// -  `7`: `MintTo`
+/// -  `9`: `CloseAccount`
+/// - `16`: `InitializeAccount2`
 /// - `18`: `InitializeAccount3`
 /// - `20`: `InitializeMint2`
 #[inline(always)]
-pub fn process_instruction(
-    _program_id: &Pubkey,
+pub(crate) fn inner_process_instruction(
     accounts: &[AccountInfo],
     instruction_data: &[u8],
 ) -> ProgramResult {
-    let (discriminator, instruction_data) = instruction_data
-        .split_first()
-        .ok_or(ProgramError::InvalidInstructionData)?;
-    let instruction = TokenInstruction::try_from(*discriminator)?;
```

Comment:
Was the usage of `TokenInstruction` supposed to be reverted?

Reply:
Yes, using the `TokenInstruction` enum increases `20` CUs at least once I added the `Batch` variant. I could not find a way to make it efficient.